spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestPartitionedWritesAsSelect.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS %s", targetTable);
+  }
+
+  @Test
+  public void testInsertAsSelectAppend() {
+    insertData(3);
+    List<Object[]> expected = currentData();
+
+    sql("CREATE TABLE %s (id bigint, data string, category string, ts timestamp)" +
+        "USING iceberg PARTITIONED BY (days(ts), category)", targetTable);
+
+    sql("INSERT INTO %s SELECT id, data, category, ts FROM %s ORDER BY ts,category", targetTable, tableName);
+    Assert.assertEquals("Should have 24 rows after insert",
+        3 * 8L, scalarSql("SELECT count(*) FROM %s", targetTable));
+
+    assertEquals("Row data should match expected",
+        expected, sql("SELECT * FROM %s ORDER BY id", targetTable));
+  }
+
+  @Test
+  public void testInsertAsSelectWithBucket() {
+    insertData(3);
+    List<Object[]> expected = currentData();
+
+    sql("CREATE TABLE %s (id bigint, data string, category string, ts timestamp)" +
+        "USING iceberg PARTITIONED BY (bucket(8, data))", targetTable);
+
+    IcebergSpark.registerBucketUDF(spark, "iceberg_bucket8", DataTypes.StringType, 8);
+    sql("INSERT INTO %s SELECT id, data, category, ts FROM %s ORDER BY iceberg_bucket8(data)", targetTable, tableName);
+    Assert.assertEquals("Should have 24 rows after insert",
+        3 * 8L, scalarSql("SELECT count(*) FROM %s", targetTable));
+
+    assertEquals("Row data should match expected",
+        expected, sql("SELECT * FROM %s ORDER BY id", targetTable));
+  }
+
+  @Test
+  public void testInsertAsSelectWithTruncate() {
+    insertData(3);
+    List<Object[]> expected = currentData();
+
+    sql("CREATE TABLE %s (id bigint, data string, category string, ts timestamp)" +
+        "USING iceberg PARTITIONED BY (truncate(data, 4), truncate(id, 4))", targetTable);
+
+    IcebergSpark.registerTruncateUDF(spark, "iceberg_truncate_string4", DataTypes.StringType, 4);
+    IcebergSpark.registerTruncateUDF(spark, "iceberg_truncate_long4", DataTypes.LongType, 4);
+    sql("INSERT INTO %s SELECT id, data, category, ts FROM %s " +
+        "ORDER BY iceberg_truncate_string4(data),iceberg_truncate_long4(id)", targetTable, tableName);
+    Assert.assertEquals("Should have 24 rows after insert",
+        3 * 8L, scalarSql("SELECT count(*) FROM %s", targetTable));
+
+    assertEquals("Row data should match expected",
+        expected, sql("SELECT * FROM %s ORDER BY id", targetTable));
+  }
+
+  @Test
+  public void testInsertAsSelectWithBucketFailure() {
+    insertData(3);
+
+    sql("CREATE TABLE %s (id bigint, data string, category string, ts timestamp)" +
+        "USING iceberg PARTITIONED BY (bucket(8, id))", targetTable);
+    AssertHelpers.assertThrows("should registered temporary function" +
+        "nor a permanent function registered in the database 'default'",
+        AnalysisException.class,
+        "Undefined function: 'iceberg_bucket8'",
```

Comment:
I don't think it is necessary to test the case where you reference a function that has not been registered.

Reply:
Thanks for your review.  This code is really redundant